### DISCUSSION TITLE
Piccolo ORM outside of multi-app structure.

### DISCUSSION THREAD

**Discussion by gnat** [piccolo-orm/piccolo#619] [Ideas]
It's a bit of a hangup to use Piccolo in a multi-app layout, many ASGI servers support having a single `app.py` file you can just point to and serve instantly (Sanic Server, Uvicorn, etc).

It would be really great if Piccolo had better support to load config from any variable you choose, and more flexibility about where it can find config (not just loaded from modules)- so you can do migrations as a single app.

## A bit of a hack, but I have it working like this:

* /project
  * app.py
  * /tables
    * tables.py 
    * /migrations
    
```python 
import os
from piccolo.conf.apps import AppRegistry
from piccolo.engine.postgres import PostgresEngine
from piccolo.conf.apps import AppConfig, table_finder

class AppRegistryLocal(AppRegistry):
    def __init__(self, local):
        app_config = local
        self.apps = [app_config.app_name]
        self.app_configs = {}
        app_names = []
        self.app_configs[app_config.app_name] = app_config
        app_names.append(app_config.app_name)
        self._validate_app_names(app_names)

CURRENT_DIRECTORY = os.path.dirname(os.path.abspath(__file__))

APP_CONFIG = AppConfig(
    app_name="app",
    migrations_folder_path=os.path.join(
        CURRENT_DIRECTORY, "tables", "migrations"
    ),
    table_classes=table_finder(modules=["tables.tables"]),
    migration_dependencies=[],
    commands=[],
)

DB = PostgresEngine(config={
    "database": "piccolo",
    "user": "root",
    "password": "",
    "host": "localhost",
    "port": 26257,
})

APP_REGISTRY = AppRegistryLocal(local=APP_CONFIG)
```

Using this layout, piccolo can be run like so:
* `export PICCOLO_CONF=app; piccolo migrations new app --auto`
* `export PICCOLO_CONF=app; piccolo migrations forwards app`

**Comment by gnat** [comment]
Thoughts? More intuitive patterns?

This would be especially convenient for teams wanting to migrate an existing app to Piccolo without restructuring- fewer barriers to adoption.

Would be great to at least see AppRegistry be able to take configuration from any variable before 1.0.

**Comment by gnat** [comment]
Note, when you run your web server you also have to use `export PICCOLO_CONF=app;` so `Finder` will work... rough edge that may need some smoothing out for this to work well.

**Comment by dantownsend** [comment]
The easiest hack would be to put all of your code in `piccolo_conf.py`.

What makes things awkward at the moment is if you wanted to rename `piccolo_conf.py` to `app.py` then you have to use the `PICCOLO_CONF=app` environment variable to tell Piccolo where to look.

There could be another way besides environment variables - a `.piccolo` file or something.
